crates/subspace-networking/src/peer_info/handler.rs

```diff
@@ -0,0 +1,281 @@
+use crate::peer_info::{protocol, PeerInfo};
+use futures::future::BoxFuture;
+use futures::prelude::*;
+use libp2p::core::upgrade::{NegotiationError, ReadyUpgrade};
+use libp2p::core::UpgradeError;
+use libp2p::swarm::handler::{
+    ConnectionEvent, DialUpgradeError, FullyNegotiatedInbound, FullyNegotiatedOutbound,
+    ListenUpgradeError,
+};
+use libp2p::swarm::{
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, KeepAlive,
+    NegotiatedSubstream, SubstreamProtocol,
+};
+use std::error::Error;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use std::{fmt, io};
+use tracing::debug;
+
+/// The configuration for peer-info protocol.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// Protocol timeout.
+    timeout: Duration,
+
+    /// Protocol name.
+    protocol_name: &'static [u8],
+}
+
+impl Config {
+    /// Creates a new [`Config`] with the following default settings:
+    ///
+    ///   * [`Config::with_timeout`] 20s
+    pub fn new(protocol_name: &'static [u8]) -> Self {
+        Self {
+            timeout: Duration::from_secs(20),
+            protocol_name,
+        }
+    }
+
+    /// Sets the protocol timeout.
+    pub fn with_timeout(mut self, d: Duration) -> Self {
+        self.timeout = d;
+        self
+    }
+}
+
+/// The successful result of processing an inbound or outbound peer info requests.
+#[derive(Debug)]
+pub enum PeerInfoSuccess {
+    Received(PeerInfo),
+    DataSent,
+}
+
+/// A peer info protocol failure.
+#[derive(Debug)]
+pub enum PeerInfoError {
+    /// The peer does not support the peer info protocol.
+    Unsupported,
+    /// The peer info request failed.
+    Other {
+        error: Box<dyn Error + Send + 'static>,
+    },
+}
+
+impl fmt::Display for PeerInfoError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            PeerInfoError::Other { error } => write!(f, "Peer info error: {error}"),
+            PeerInfoError::Unsupported => write!(f, "Peer info protocol not supported"),
+        }
+    }
+}
+
+impl Error for PeerInfoError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            PeerInfoError::Other { error } => Some(&**error),
+            PeerInfoError::Unsupported => None,
+        }
+    }
+}
```

Comment:
Better use `thiserror`, then you can just write `#[source]` and `Display` will be derived for you